### PR TITLE
Addresses #193 & #196 (cli notifies when nothing to download/watch on seasonals) +more

### PR DESCRIPTION
**Pull request by DragonOfShuu**
This change addresses #193 and #196 by finding if the user has nothing new to download/watch, and notifies accordingly, instead of immediately crashing.

Not only that, but with the API you can now retry the download multiple times determined by a limit by the programmer. I implemented this because while I attempted to download anime, every 3 animes it errored while downloading, but I found if the API just retried, it would work again. 

However, if retrying doesn't work, the API will just re-raise the error, meaning the programmer using the API can handle it from there. The CLI will now tell the user if there was an issue downloading, and will either skip downloading that episode, or it will just stop downloading that entire series (depending on the mode). 

# TLDR;

- User notified in seasonal/mal mode if there is nothing to watch/download
- API can retry downloads multiple times
- CLI catches download errors and properly notifies the user
- CLI has a new DownloadComp, which is flexible enough to be used across different menus/clis

### CODE REVIEW

File: cli/src/anipy_cli/menus/seasonal_menu.py [hunk 1]
```diff
@@ -180,65 +186,37 @@ def change_lang(self):
...
-                        colors.END,
-                        " Episode ",
-                        ep,
-                        "...",
-                    )
 
-                    stream = anime.get_video(
-                        ep, lang, preferred_quality=self.options.quality
-                    )
+        def onSuccessfulDownload(anime: Anime, ep: Episode, lang: LanguageTypeEnum):
```
Comment (sdaqo) [inline cli/src/anipy_cli/menus/seasonal_menu.py:206]: snake case here as well

File: cli/src/anipy_cli/menus/seasonal_menu.py [hunk 2]
```diff
@@ -180,65 +186,37 @@ def change_lang(self):
...
-                        container=config.remux_to,
-                        ffmpeg=self.options.ffmpeg or config.ffmpeg_hls,
-                    )
-                    self.seasonal_list.update(anime, episode=ep, language=lang)
+        failed_series = DownloadComponent(self.options, self.dl_path).download_anime(picked, onSuccessfulDownload)
 
         if not self.options.auto_update:
-            self.print_options(clear_screen=True)
+            # Clear screen only if there were no issues
+            self.print_options(clear_screen=len(failed_series)==0)
```
Comment (sdaqo) [inline cli/src/anipy_cli/menus/seasonal_menu.py:213]: nit: proper spacing here
```suggestion
            self.print_options(clear_screen=len(failed_series) == 0)
```

File: cli/src/anipy_cli/menus/mal_menu.py [hunk 3]
```diff
@@ -249,57 +251,28 @@ def tag_anime(self):
...
         if total_eps == 0:
             print("Nothing to download, returning...")
             return
         else:
             print(f"Downloading a total of {total_eps} episode(s)")
+        
+        convert: Dict[Anime, MALAnime] = {d[0]: d[1] for d in picked}
+        new_picked = [(anime_info[0], anime_info[2], anime_info[3]) for anime_info in picked]
+
+        def onSuccessDownload(anime: Anime, ep: Episode, _: LanguageTypeEnum):
```
Comment (sdaqo) [inline cli/src/anipy_cli/menus/mal_menu.py:264]:
```suggestion
        def on_success_download(anime: Anime, ep: Episode, _: LanguageTypeEnum):
```

File: cli/src/anipy_cli/menus/mal_menu.py [hunk 4]
```diff
@@ -249,57 +251,28 @@ def tag_anime(self):
...
-                        container=config.remux_to,
-                        ffmpeg=self.options.ffmpeg or config.ffmpeg_hls,
-                    )
-                    if not all:
-                        self.mal_proxy.update_show(
-                            mal_anime,
-                            status=MALMyListStatusEnum.WATCHING,
-                            episode=int(ep),
-                        )
+        self.print_options(clear_screen=len(errors)==0)
```
Comment (sdaqo) [inline cli/src/anipy_cli/menus/mal_menu.py:275]:
```suggestion
        self.print_options(clear_screen=len(errors) == 0)
```

File: cli/src/anipy_cli/comps/DownloadComp.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        self.dl_path = dl_path
+
+    def download_anime(
+        self,
+        picked: List[Tuple[Anime, LanguageTypeEnum, List[Episode]]],
+        /,
+        after_success_ep: Callable[[Anime, Episode, LanguageTypeEnum], Any] = lambda _,__,___: None,
+        only_skip_ep_on_err: bool = False,
+    ) -> List[Tuple[Anime, Episode]]:
+        '''
```
Comment (sdaqo) [inline cli/src/anipy_cli/comps/DownloadComp.py:30]: please use double quotes for the docstrings

File: cli/src/anipy_cli/comps/DownloadComp.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    """
+    def __init__(self, cliArgs: CliArgs, dl_path: Path) -> None:
+        self.options = cliArgs
+        self.dl_path = dl_path
+
+    def download_anime(
+        self,
+        picked: List[Tuple[Anime, LanguageTypeEnum, List[Episode]]],
+        /,
+        after_success_ep: Callable[[Anime, Episode, LanguageTypeEnum], Any] = lambda _,__,___: None,
```
Comment (sdaqo) [inline cli/src/anipy_cli/comps/DownloadComp.py:27]: please use a protocol for the callback as done in for example in the download file in the api

File: cli/src/anipy_cli/comps/DownloadComp.py
Comment (sdaqo) [inline cli/src/anipy_cli/comps/DownloadComp.py:1]: please rename this file to fit with the snake case names of the other files, also there is no need for this to be in a separate directory (only one file), just move it in the root

File: cli/src/anipy_cli/comps/DownloadComp.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+                    eps,
+                    after_success_ep,
+                    only_skip_ep_on_err,
+                )
+
+            return failed
+
+    def download_episodes(
+        self,
+        s: DotSpinner,
```
Comment (sdaqo) [inline cli/src/anipy_cli/comps/DownloadComp.py:66]: please use the full name "spinner" for the argument

File: cli/src/anipy_cli/comps/DownloadComp.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                if (only_skip_ep_on_err):
+                    continue
+                return fails
+            
+            after_success_ep(anime, ep, lang)
+        return fails
+
+    def download_ep(
+        self,
+        s: DotSpinner,
```
Comment (sdaqo) [inline cli/src/anipy_cli/comps/DownloadComp.py:101]: same as above, "spinner"

File: cli/src/anipy_cli/comps/DownloadComp.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    the ani-py CLI.
+    """
+    def __init__(self, cliArgs: CliArgs, dl_path: Path) -> None:
+        self.options = cliArgs
+        self.dl_path = dl_path
+
+    def download_anime(
+        self,
+        picked: List[Tuple[Anime, LanguageTypeEnum, List[Episode]]],
+        /,
```
Comment (sdaqo) [inline cli/src/anipy_cli/comps/DownloadComp.py:26]: are those needed?

**Review by sdaqo — CHANGES_REQUESTED** [review]
Just some minor changes and nitpicks! Thanks for this!

**Comment by DragonOfShuu** [comment]
Oh no! I agree with nearly all of these changes; it clearly got wayy to late last night 💀. I'll go through and fix these, and see if I can spot any other potential issues

**Comment by sdaqo** [comment]
Haha, no worries, this is absolutely relatable

File: cli/src/anipy_cli/comps/DownloadComp.py
Comment (DragonOfShuu) [reply cli/src/anipy_cli/comps/DownloadComp.py:1]: I'm renaming the file; however, I think we may want to consider keeping the directory. I think we need to componentize more of the CLI in a similar way

**Comment by DragonOfShuu** [comment]
I went ahead and fixed all the issues outlined + more. However, I kept the download_component in a separate directory because I feel we have more parts of the CLI we can componentize

File: cli/src/anipy_cli/comps/DownloadComp.py
Comment (sdaqo) [reply cli/src/anipy_cli/comps/DownloadComp.py:1]: Considering the discord and mal code, which are in the same vein, basically a big util not shoved in the utils file, are already in the root I think for now we should move it up. This could be tackled another time finishing all at once.

Comment (sdaqo) [reply cli/src/anipy_cli/comps/DownloadComp.py:1]: If you wanna do that, please move it to another PR so this one does not get too big.

Comment (DragonOfShuu) [reply cli/src/anipy_cli/comps/DownloadComp.py:1]: That's fair. I'll move that file to the root!

**Comment by DragonOfShuu** [comment]
I have now moved the download component into the root dir!

**Review by sdaqo — APPROVED** [review]
Alright, LGTM
